x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportForecastJobAction.java

```diff
@@ -124,6 +139,26 @@ private void getForecastRequestStats(String jobId, String forecastId, ActionList
         jobResultsProvider.getForecastRequestStats(jobId, forecastId, forecastRequestStatsHandler, listener::onFailure);
     }
 
+    static Long getAdjustedMemoryLimit(Job job, Long requestedLimit, AbstractAuditor<? extends AbstractAuditMessage> auditor) {
+        if (requestedLimit == null || requestedLimit == DEFAULT_MAX_MODEL_MEMORY.getBytes()) {
+            return null;
+        }
+        long jobLimitMegaBytes = job.getAnalysisLimits() == null || job.getAnalysisLimits().getModelMemoryLimit() == null ?
+            AnalysisLimits.DEFAULT_MODEL_MEMORY_LIMIT_MB :
+            job.getAnalysisLimits().getModelMemoryLimit();
+        long allowedMax = (long)(new ByteSizeValue(jobLimitMegaBytes, ByteSizeUnit.MB).getBytes() * 0.40);
+        long adjustedMax = Math.min(requestedLimit, allowedMax - 1);
```

Comment:
Why subtract 1 from `allowedMax`?

Reply:
@droberts195 the check on the c++ side is `>=`. This is to make it strictly less than.

Reply:
This is going to lead to confusion when the numbers are on the boundary.  For example, suppose the job model memory limit is 200MB, and somebody configures 80MB for a forecast.  Then we'll create an audit message saying the requested amount _exceeded_ 40% of the job model memory limit.  But it doesn't - it's exactly 40%.

Probably the best way around this would be to change the check on the C++ side to be `>` so that round numbers work out nicely.  Or alternatively change the error message below to say "must be less than 40%".